### PR TITLE
Add source type 2153

### PR DESCRIPTION
**Pull request by TadeuRodrigues**
New pull request replacing  #2383

### Description (*)
Added source type attribute in form of source.

### Fixed Issues (if relevant)
1. magento/inventory#2153:  Add Source Type extension attribute

### Manual testing scenarios (*)
1. add or edit some source
2. you can see a new field about source's type 

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: Inventory/Model/ResourceModel/SourceType.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+declare(strict_types=1);
+
+namespace Magento\Inventory\Model\ResourceModel;
+
+use Magento\Framework\Model\ResourceModel\Db\AbstractDb;
+
+/**
+ * Implementation of basic operations for Source Type entity for specific db layer
+ */
+class SourceType extends AbstractDb
```
Comment (sidolov) [inline Inventory/Model/ResourceModel/SourceType.php]: Looks like the class is redundant, it's copy of `Inventory/Model/ResourceModel/Source/Type.php`

File: Inventory/Model/GetSourceTypeBySourceCode.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+declare(strict_types=1);
+
+namespace Magento\Inventory\Model;
+
+use Magento\InventoryApi\Model\GetSourceTypeBySourceCodeInterface;
+
+/**
+ * @inheritdoc
+ */
+class GetSourceTypeBySourceCode implements GetSourceTypeBySourceCodeInterface
```
Comment (novikor) [inline Inventory/Model/GetSourceTypeBySourceCode.php]: Can we remove this class by using `Magento\Inventory\Model\ResourceModel\GetSourceTypeBySourceCode` directly?

File: Inventory/Model/ResourceModel/GetSourceTypeBySourceCode.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        $tableName = $this->resourceConnection->getTableName(SourceTypeLink::TABLE_NAME_SOURCE_TYPE_LINK);
+        $connection = $this->resourceConnection->getConnection();
+
+        $qry = $connection
+            ->select()
+            ->from($tableName, 'type_code')
+            ->where('source_code = ?', $sourceCode);
+
+        return $connection->fetchOne($qry);
```
Comment (novikor) [inline Inventory/Model/ResourceModel/GetSourceTypeBySourceCode.php:46]: What happens if the select returns nothing?
It seems that `NoSuchEntityException` should be used there

File: Inventory/Model/ResourceModel/Source/Type.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    public function _construct()
+    {
+        $this->_init('inventory_source_type', 'type_code');
+    }
+
+    /**
+     * Return all types of source to select field
+     *
+     * @return array
+     * @throws \Magento\Framework\Exception\LocalizedException
```
Comment (novikor) [inline Inventory/Model/ResourceModel/Source/Type.php]: May be replaced with import

File: Inventory/Model/ResourceModel/Source/Type.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    public function getAllTypes()
```
Comment (novikor) [inline Inventory/Model/ResourceModel/Source/Type.php]:
```suggestion
    public function getAllTypes(): array
```

File: Inventory/Model/ResourceModel/SourceTypeLink.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * Implementation of basic operations for SourceTypeLink entity for specific db layer
+ *
+ * This class needed for internal purposes only, to make collection work properly
+ */
+class SourceTypeLink extends AbstractDb
+{
+    /**#@+
+     * Constants related to specific db layer
+     */
+    const TABLE_NAME_SOURCE_TYPE_LINK = 'inventory_source_type_link';
```
Comment (novikor) [inline Inventory/Model/ResourceModel/SourceTypeLink.php:22]: It might be better to use private constants

File: Inventory/Model/ResourceModel/TypeSource/Collection.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    /**
+     * Convert collection items to select options array
+     *
+     * @return array
+     */
+    public function toOptionArray()
+    {
+
```
Comment (novikor) [inline Inventory/Model/ResourceModel/TypeSource/Collection.php]:
```suggestion
```

File: Inventory/Model/SourceTypeLink/Command/SourceTypeLinkDelete.php [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        if (empty($sourceCode)) {
+            throw new InputException(__('Input data is empty'));
+        }
+
+        try {
+            $this->delete->execute($sourceCode);
+        } catch (\Exception $e) {
+            $this->logger->error($e->getMessage());
+            throw new CouldNotDeleteException(__('Could not delete SourceTypeLink'), $e);
```
Comment (novikor) [inline Inventory/Model/SourceTypeLink/Command/SourceTypeLinkDelete.php:56]: The interface has missing `@throws` tags:
image

File: Inventory/Model/SourceTypeLink/Command/SourceTypeLinkSave.php [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    public function execute(SourceTypeLinkInterface $link): void
+    {
+        if (empty($link)) {
+            throw new InputException(__('Input data is empty'));
+        }
+
+        $validationResult = $this->sourceTypeLinkValidator->validate($link);
+        if (!$validationResult->isValid()) {
+            throw new ValidationException(__('Validation Failed'), null, 0, $validationResult);
```
Comment (novikor) [inline Inventory/Model/SourceTypeLink/Command/SourceTypeLinkSave.php:66]: image

Please add `@throws` tag to the interface

File: Inventory/Model/SourceTypeLink/Validator/SourceCodeValidator.php [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+use Magento\Inventory\Model\Validators\NoSpecialCharsInString;
+use Magento\Inventory\Model\Validators\NotAnEmptyString;
+use Magento\Inventory\Model\Validators\NoWhitespaceInString;
+
+/**
+ * Check that code is valid
+ */
+class SourceCodeValidator implements SourceTypeLinkValidatorInterface
+{
+
```
Comment (novikor) [inline Inventory/Model/SourceTypeLink/Validator/SourceCodeValidator.php:23]:
```suggestion
```

File: Inventory/Model/SourceTypeLink/Validator/TypeCodeValidator.php [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+use Magento\Inventory\Model\Validators\NoSpecialCharsInString;
+use Magento\Inventory\Model\Validators\NotAnEmptyString;
+use Magento\Inventory\Model\Validators\NoWhitespaceInString;
+
+/**
+ * Check that code is valid
+ */
+class TypeCodeValidator implements SourceTypeLinkValidatorInterface
+{
+
```
Comment (novikor) [inline Inventory/Model/SourceTypeLink/Validator/TypeCodeValidator.php:23]:
```suggestion
```

File: Inventory/Observer/AfterSourceSave.php [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+use Magento\InventoryApi\Api\Data\SourceTypeLinkInterfaceFactory;
+use Magento\InventoryApi\Api\SourceTypeLinkDeleteInterface;
+use Magento\InventoryApi\Api\SourceTypeLinkSaveInterface;
+
+/**
+ * Plugin to save type of source after save source
+ */
+class AfterSourceSave implements ObserverInterface
+{
+
```
Comment (novikor) [inline Inventory/Observer/AfterSourceSave.php]:
```suggestion
```

File: Inventory/Observer/AfterSourceSave.php [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    private function saveTypeLinksBySource(SourceInterface $source): void
+    {
+        $linkData = [
+            'source_code' => $source->getSourceCode(),
+            'type_code' => $source->getExtensionAttributes()->getTypeCode()
+        ];
+
+        /** @var SourceTypeLinkInterface $link */
+        $link = $this->sourceTypeLinkFactory->create();
+        $link->addData($linkData);
```
Comment (novikor) [inline Inventory/Observer/AfterSourceSave.php]: It would be better to use `Magento\Framework\Api\DataObjectHelper::populateWithArray` instead of DataObject methods (`addData`)

File: Inventory/Observer/AfterSourceSave.php [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        $this->sourceTypeLinkFactory = $sourceTypeLinkFactory;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function execute(Observer $observer)
+    {
+        /** @var SourceInterface $source */
+        $source = $observer->getEvent()->getSource();
```
Comment (novikor) [inline Inventory/Observer/AfterSourceSave.php:72]: Well, there will not be a case when a `$source` is null, but please add the corresponding check anyway)

File: InventoryAdminUi/Model/OptionSource/TypeSource.php [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+use Magento\Framework\Data\OptionSourceInterface;
+use Magento\Inventory\Model\ResourceModel\TypeSource\Collection;
+use Magento\Inventory\Model\ResourceModel\TypeSource\CollectionFactory;
+
+/**
+ * Provide option values for UI
+ */
+class TypeSource implements OptionSourceInterface
+{
+
```
Comment (novikor) [inline InventoryAdminUi/Model/OptionSource/TypeSource.php:19]:
```suggestion
```

File: InventoryAdminUi/Ui/DataProvider/SourceDataProvider.php [hunk 16]
```diff
@@ -116,10 +116,15 @@ public function getData()
             if ($data['totalRecords'] > 0) {
                 $sourceCode = $data['items'][0][SourceInterface::SOURCE_CODE];
                 $sourceGeneralData = $data['items'][0];
+
+                $source = $this->sourceRepository->get($sourceCode);
+                $sourceGeneralData['type_code'] = $source->getExtensionAttributes()->getTypeCode();
```
Comment (novikor) [inline InventoryAdminUi/Ui/DataProvider/SourceDataProvider.php]: Which problem did you try to solve?  Empty `extension_attributes` or just having the `type_code` array key?
I think that we can revert these changes and use the extension attribute this way (just using another `dataScope`):  

https://github.com/magento/inventory/blob/62f89e2f13f776cc6d30f97d4f2922e734ed3207/InventoryInStorePickupAdminUi/view/adminhtml/ui_component/inventory_source_form.xml#L71

File: InventoryAdminUi/view/adminhtml/ui_component/inventory_source_form.xml [hunk 17]
```diff
@@ -94,6 +94,22 @@
                 </checkbox>
             </formElements>
         </field>
+        <field name="type_code" formElement="select" sortOrder="30">
```
Comment (novikor) [inline InventoryAdminUi/view/adminhtml/ui_component/inventory_source_form.xml:97]: https://github.com/magento/inventory/blob/62f89e2f13f776cc6d30f97d4f2922e734ed3207/InventoryInStorePickupAdminUi/view/adminhtml/ui_component/inventory_source_form.xml#L71

It would be better to use something like `extension_attributes.type_code` (after reverting the `SourceDataProvider`)

File: InventoryAdminUi/view/adminhtml/ui_component/inventory_source_listing.xml [hunk 18]
```diff
@@ -134,6 +134,18 @@
                 </editor>
             </settings>
         </column>
+        <column name="type_code" sortOrder="35">
```
Comment (novikor) [inline InventoryAdminUi/view/adminhtml/ui_component/inventory_source_listing.xml]: Should be remade after the SourceDataObject reverting.
You may see https://github.com/magento/inventory/blob/62f89e2f13f776cc6d30f97d4f2922e734ed3207/InventoryInStorePickupAdminUi/view/adminhtml/ui_component/inventory_source_listing.xml#L10 as an example

File: InventoryApi/Model/GetSourceTypeBySourceCodeInterface.php [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryApi\Model;
+
+/**
+ * Sugar service to retrieve source type by source code
+ *
```
Comment (novikor) [inline InventoryApi/Model/GetSourceTypeBySourceCodeInterface.php]:
```suggestion
```

File: InventoryApi/Model/SourceTypeLinkValidatorChain.php [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryApi\Model;
+
+use Magento\InventoryApi\Api\Data\SourceTypeLinkInterface;
+use Magento\Framework\Validation\ValidationResultFactory;
```
Comment (novikor) [inline InventoryApi/Model/SourceTypeLinkValidatorChain.php:11]: image

File: InventoryCatalog/Setup/Patch/Schema/InitializeDefaultStock.php [hunk 21]
```diff
@@ -53,7 +71,9 @@ public function __construct(
     public function apply()
     {
         $this->createDefaultSource->execute();
+        $this->createDefaultSourceTypes->execute();
```
Comment (novikor) [inline InventoryCatalog/Setup/Patch/Schema/InitializeDefaultStock.php]: Hmm. I afraid it should work for pure MSI installation but could fail in the existing MSI installation upgrade.
@sidolov, what do you think?

**Review by novikor — CHANGES_REQUESTED** [review]
Generally, fine, - good job! Please see my comments.
Also, executing 
```bash
vendor/bin/php-cs-fixer fix ext/magento/inventory/Inventory --config=.php_cs.dist 
```
Could say something interesting)

File: InventoryCatalog/Setup/Patch/Schema/InitializeDefaultStock.php (hunk 21 above)
Comment (TadeuRodrigues) [reply InventoryCatalog/Setup/Patch/Schema/InitializeDefaultStock.php]: I gonna make a new patch to create types also assign.

**Review by sidolov — CHANGES_REQUESTED** [review]

File: Inventory/Model/ResourceModel/SourceType.php (hunk 1 above)
Comment (TadeuRodrigues) [reply Inventory/Model/ResourceModel/SourceType.php]: your are right, I'm fixing now

File: Inventory/Model/ResourceModel/SourceTypeLink.php (hunk 6 above)
Comment (TadeuRodrigues) [reply Inventory/Model/ResourceModel/SourceTypeLink.php:22]: i can't because I use lot of time:
[Inventory/Model/ResourceModel/SourceTypeLink/Save.php](https://github.com/TadeuRodrigues/msi/blob/92ca7315979f58908d2af7ac1dddf951898161c5/Inventory/Model/ResourceModel/SourceTypeLink/Save.php#L53)

[Inventory/Model/ResourceModel/SourceTypeLink/Delete.php](https://github.com/TadeuRodrigues/msi/blob/92ca7315979f58908d2af7ac1dddf951898161c5/Inventory/Model/ResourceModel/SourceTypeLink/Delete.php#L43)

[Inventory/Model/ResourceModel/GetSourceTypeBySourceCode.php](https://github.com/TadeuRodrigues/msi/blob/92ca7315979f58908d2af7ac1dddf951898161c5/Inventory/Model/ResourceModel/GetSourceTypeBySourceCode.php#L38)

File: Inventory/Model/GetSourceTypeBySourceCode.php (hunk 2 above)
Comment (TadeuRodrigues) [reply Inventory/Model/GetSourceTypeBySourceCode.php]: removed

File: Inventory/Model/Source/InitTypeCodeExtensionAttributes.php [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+    public function __construct(
+        ExtensionAttributesFactory $extensionAttributesFactory,
+        GetSourceTypeBySourceCode $getSourceTypeBySourceCode
+    ) {
+        $this->extensionAttributesFactory = $extensionAttributesFactory;
+        $this->getSourceTypeBySourceCode = $getSourceTypeBySourceCode;
+    }
+
+    /**
+     * Set store-pickup related source extension attributes.
```
Comment (novikor) [inline Inventory/Model/Source/InitTypeCodeExtensionAttributes.php]: Please remove the "store-pickup" from phpdocs :)

**Review by novikor — COMMENTED** [review]
@sidolov, can we modify a source repository/collection directly in order to load the extension attributes instead of creating plugins? In our case, we are operating in the scope of the same module.

File: Inventory/Model/Source/InitTypeCodeExtensionAttributes.php (hunk 22 above)
Comment (TadeuRodrigues) [reply Inventory/Model/Source/InitTypeCodeExtensionAttributes.php:44]: Exacly, i was thinking about this too, we can save lot of process with this change.

**Comment by TadeuRodrigues** [comment]
@novikor exactly, I was thinking about this too, we can save lot of process with this change.

**Comment by sidolov** [comment]
Hi @TadeuRodrigues , according to the priorities current issue should be delivered to 1.2 release line only. I'm closing the PR since it will not be delivered to the 1.1 release line. Feel free to open PR to 1.2 release line. Thank you for your contribution!
